File: src/mocks/statistics.ts
import type {
  AddedCardsStats,
  CardBreakdown,
  CardCategory,
  CardCategoryCount,
  DailyCount,
  ReviewHistoryStats,
  TimeRange,
  TodayProgress,
} from '../types/statistics'

import type { AnkiMock } from './anki'
import type { Card } from '../types/anki'

/** 历史数据覆盖的天数。 */
const HISTORY_DAYS = 365

const CATEGORY_LABELS: Record<CardCategory, string> = {
  new: '新卡',
  learning: '学习中',
  review: '复习中',
  relearning: '重新学习',
}

function localDay(date: Date): string {
  const month = `${date.getMonth() + 1}`.padStart(2, '0')
  const day = `${date.getDate()}`.padStart(2, '0')
  return `${date.getFullYear()}-${month}-${day}`
}

function dayOffset(days: number): string {
  const date = new Date()
  date.setDate(date.getDate() + days)
  return localDay(date)
}

/** 固定种子的线性同余发生器：保证每次刷新看到相同的历史曲线。 */
function createRandom(seed: number) {
  let state = seed
  return () => {
    state = (state * 1664525 + 1013904223) % 4294967296
    return state / 4294967296
  }
}

/** 生成近一年的每日计数（最旧在前）。 */
function seedDailyCounts(seed: number, maxPerDay: number, idleChance: number): number[] {
  const random = createRandom(seed)
  const counts: number[] = []
  for (let index = 0; index < HISTORY_DAYS; index += 1) {
    counts.push(random() < idleChance ? 0 : Math.floor(random() * maxPerDay) + 1)
  }
  return counts
}

/** 内存版统计后端：卡片分布取自 AnkiMock，历史曲线为固定种子的假数据。 */
export class StatisticsMock {
  private reviewCounts = seedDailyCounts(20260101, 12, 0.55)
  private addedCounts = seedDailyCounts(20260917, 5, 0.82)
  private reviewedToday = new Set<string>()

  constructor(private anki: AnkiMock) {
    anki.setReviewListener((cardId) => {
      this.reviewedToday.add(cardId)
      this.reviewCounts[HISTORY_DAYS - 1] += 1
    })
  }

  handle(cmd: string, payload: Record<string, unknown>): unknown {
    switch (cmd) {
      case 'stats_get_today_progress':
        return this.todayProgress()
      case 'stats_get_card_breakdown':
        return this.breakdown()
      case 'stats_get_review_history':
        return this.reviewHistory(String(payload.range ?? 'last_year') as TimeRange)
      case 'stats_get_added_cards':
        return this.addedCards(String(payload.range ?? 'last_year') as TimeRange)
      default:
        return undefined
    }
  }

  private cards(): Card[] {
    return this.anki.snapshotCards()
  }

  private todayProgress(): TodayProgress {
    const now = Date.now()
    const cards = this.cards()
    const newRemaining = cards.filter((card) => card.state === 'new').length
    const dueRemaining = cards.filter(
      (card) =>
        card.state !== 'new' &&
        card.dueAt !== null &&
        new Date(card.dueAt).getTime() <= now,
    ).length

    const pendingCards = newRemaining + dueRemaining
    const reviewedCards = this.reviewedToday.size
    const answered = reviewedCards + pendingCards

    return {
      date: localDay(new Date()),
      reviewedCards,
      pendingCards,
      totalCards: this.cards().length,
      newRemaining,
      dueRemaining,
      completionPercent: answered === 0
        ? 0
        : Math.round((reviewedCards / answered) * 1000) / 10,
    }
  }

  private breakdown(): CardBreakdown {
    const buckets = new Map<CardCategory, number>()
    const add = (category: CardCategory) =>
      buckets.set(category, (buckets.get(category) ?? 0) + 1)

    for (const card of this.cards()) {
      add(card.state)
    }

    const categories = (Object.keys(CATEGORY_LABELS) as CardCategory[])
      .map<CardCategoryCount>((category) => ({
        category,
        label: CATEGORY_LABELS[category],
        count: buckets.get(category) ?? 0,
        percent: 0,
      }))

    const total = categories.reduce((sum, item) => sum + item.count, 0)
    for (const item of categories) {
      item.percent = total === 0 ? 0 : Math.round((item.count / total) * 1000) / 10
    }

    return { total, categories }
  }

  private reviewHistory(range: TimeRange): ReviewHistoryStats {
    const days = this.slice(range, this.reviewCounts)
    const totalReviews = days.reduce((sum, day) => sum + day.count, 0)
    const studiedDays = days.filter((day) => day.count > 0).length
    const elapsedDays = days.length

    return {
      range,
      days,
      totalReviews,
      studiedDays,
      elapsedDays,
      studiedDayPercent: elapsedDays === 0
        ? 0
        : Math.round((studiedDays / elapsedDays) * 1000) / 10,
      averagePerElapsedDay: elapsedDays === 0
        ? 0
        : Math.round((totalReviews / elapsedDays) * 100) / 100,
      averagePerStudiedDay: studiedDays === 0
        ? 0
        : Math.round((totalReviews / studiedDays) * 100) / 100,
    }
  }

  private addedCards(range: TimeRange): AddedCardsStats {
    const days = this.slice(range, this.withRealCards(this.addedCounts))
    const total = days.reduce((sum, day) => sum + day.count, 0)

    return {
      range,
      days,
      total,
      elapsedDays: days.length,
      averagePerDay: days.length === 0 ? 0 : Math.round((total / days.length) * 100) / 100,
    }
  }

  /** 把真实存在的卡片按其创建日计入新增曲线，今天新建的卡片会立即反映出来。 */
  private withRealCards(counts: number[]): number[] {
    const merged = [...counts]
    for (const card of this.cards()) {
      const day = localDay(new Date(card.createdAt))
      for (let index = 0; index < HISTORY_DAYS; index += 1) {
        if (dayOffset(index - HISTORY_DAYS + 1) === day) {
          merged[index] += 1
          break
        }
      }
    }
    return merged
  }

  private slice(range: TimeRange, counts: number[]): DailyCount[] {
    const size = range === 'last_month'
      ? 30
      : range === 'last_three_months'
        ? 90
        : HISTORY_DAYS
    const days: DailyCount[] = []

    for (let index = HISTORY_DAYS - size; index < HISTORY_DAYS; index += 1) {
      days.push({
        day: dayOffset(index - HISTORY_DAYS + 1),
        daysAgo: index - HISTORY_DAYS + 1,
        count: counts[index] ?? 0,
      })
    }
    return days
  }
}

File: src-tauri/src/interface/statistics.rs
//! Anki 卡片统计的接口类型。
//!
//! 统计只读取 `card` 与 `review_log` 两张表，所有「天」均按本地时区日期
//! （`YYYY-MM-DD`）切分，与 Anki 的统计口径保持一致。

use serde::{Deserialize, Serialize};

/// 图表时间范围。
#[derive(Serialize, Deserialize, Clone, Copy, PartialEq, Eq, Debug)]
#[serde(rename_all = "snake_case")]
pub enum TimeRange {
    /// 近 30 天（含今天）。
    LastMonth,
    /// 近 90 天（含今天）。
    LastThreeMonths,
    /// 近 365 天（含今天）。
    LastYear,
    /// 自首个有数据的日期起。
    All,
}

/// 卡片分类，与 `CardState` 一一对应，顺序即图例顺序。
#[derive(Serialize, Deserialize, Clone, Copy, PartialEq, Eq, Debug)]
#[serde(rename_all = "snake_case")]
pub enum CardCategory {
    /// 新卡。
    New,
    /// 学习中。
    Learning,
    /// 复习中。
    Review,
    /// 重新学习。
    Relearning,
}

/// 今日学习统计：首页环形图与统计页第一张卡片共用。
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct TodayProgress {
    /// 本地日期 `YYYY-MM-DD`。
    pub date: String,
    /// 今日已作答的去重卡片数。
    pub reviewed_cards: u32,
    /// 此刻仍需作答的卡片数。
    pub pending_cards: u32,
    /// 卡片总数（含暂停与搁置）。
    pub total_cards: u32,
    /// 待作答卡片中未学习的数量。
    pub new_remaining: u32,
    /// 待作答卡片中已到期需复习的数量。
    pub due_remaining: u32,
    /// 今日完成度：已完成 /（已完成 + 待完成）。
    pub completion_percent: f64,
}

/// 单个分类的数量与占比。
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct CardCategoryCount {
    pub category: CardCategory,
    /// 展示名，如「新卡」。
    pub label: String,
    pub count: u32,
    /// 占 `CardBreakdown::total` 的百分比。
    pub percent: f64,
}

/// 卡片数量分布。
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct CardBreakdown {
    pub total: u32,
    /// 固定顺序：新卡、学习中、复习中、重新学习。
    pub categories: Vec<CardCategoryCount>,
}

/// 条形图中的一天。
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct DailyCount {
    /// 本地日期 `YYYY-MM-DD`。
    pub day: String,
    /// 距今天数：`0` 为今天，`-N` 为 N 天前。
    pub days_ago: i32,
    pub count: u32,
}

/// 复习行为历史。
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct ReviewHistoryStats {
    pub range: TimeRange,
    /// 范围内每一天，无数据的天补零。
    pub days: Vec<DailyCount>,
    pub total_reviews: u32,
    /// 学习天数：范围内有作答记录的天数。
    pub studied_days: u32,
    /// 范围内包含今天在内的自然天数。
    pub elapsed_days: u32,
    /// 学习天数占比。
    pub studied_day_percent: f64,
    /// 平均值（包含未学习天数）。
    pub average_per_elapsed_day: f64,
    /// 平均值（只计实际学习天数）。
    pub average_per_studied_day: f64,
}

/// 新增卡片趋势。
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct AddedCardsStats {
    pub range: TimeRange,
    /// 范围内每一天，无新增的天补零。
    pub days: Vec<DailyCount>,
    pub total: u32,
    /// 范围内包含今天在内的自然天数。
    pub elapsed_days: u32,
    pub average_per_day: f64,
}
